accepted/single-file/design_3_0.md

```diff
@@ -0,0 +1,136 @@
+
+## Introduction
+
+In .net core 3 implements Stage 1 (described in the [staging document](staging.md)) support for single-file apps.
+
+## Build System Interface
+
+Publishing to a single file can be triggered by adding the following property to an application's project file:
+
+```xml
+<PropertyGroup>
+    <PublishSingleFile>true</PublishSingleFile>
+</PropertyGroup>    
+```
+
+* The `PublishSingleFile` property applies to both framework dependent and self-contained publish operations.
+* The `PublishSingleFile` property applies to platform-specific builds with respect to a given runtime-identifier. The output of the build is a native binary for the specified platform.
+  When `PublishSingleFile`  is set to `true`, it is an error to leave `RuntimeIdentifier` undefined, or to set `UseAppHost` to `false`.
+* Setting the `PublishSingleFile`property causes the managed app, managed dependencies, platform-specific native dependencies, configurations, etc. (basically the contents of the publish directory when `dotnet publish` is run without setting the property) to be embedded within the native `apphost`. 
+
+By default, the symbol files are not embedded within the single-file, but remain as separate files in the publish directory. This includes both the IL `.pdb` file, and the native `.ni.pdb` / `app.guid.map` files generated by ready-to-run compiler. Setting the following property causes the symbol files to be included in the single-file.
+
+```xml
+<PropertyGroup>
+    <IncludeSymbolsInSingleFile>true</IncludeSymbolsInSingleFile>
+</PropertyGroup>
+```
+
+Certain files can be explicitly excluded from being embedded in the single-file by setting following meta-data:
+
+```xml
+<ExcludeFromSingleFile>true</ExcludeFromSingleFile>
+```
+
+For example, to place some files in the publish directory but not bundle them in the single-file:
+
+```xml
+<ItemGroup>
+    <Content Update="*.xml">
+      <CopyToPublishDirectory>PreserveNewest</CopyToPublishDirectory>
+      <ExcludeFromSingleFile>true</ExcludeFromSingleFile>
+    </Content>
+  </ItemGroup>
+```
+
+## The Bundler
+
+The bundler is a tool that embeds the managed app and its dependencies into the native `AppHost` executable, as described [here](bundler.md). The bundle layout in .net core 3.x is described below.
+
+| Bundle Layout (ver 1.0) |
+| ------------------------------------------------------------ |
+| **AppHost**<br />`(Bundle-Marker)`                               |
+| **Embedded Files**<br />`app.dll` <br />`app.deps.json`<br />`app.runtimeconfig.json`<br />`dependency.dll`<br />`...`<br /> |
+| **Bundle Header** <br />`1.0` (Version #)<br /> `#Number of Embedded Files`<br />`Bundle-ID` |
+| **Bundle Manifest**<br />For each bundled file:<br />   `Location (Offset, Size)`<br />   `Type :IL, ReadyToRun, other` (This information is not used in 3.0) |
+
+## The Host
+
+On Startup, the AppHost checks if it has embedded files. If so, it 
+
+* Memory maps the entire bundle file.
```

Comment:
The entire bundle file can be quite large, and on 32-bit systems there might be not enough virtual address to do this. There should be a fallback.